feat: resp protocol parsing for sorted set methods

anitarua opened the pull request:
Edit: adds RESP protocol request parsing for sorted set methods that will be supported in the momento proxy, which was recently forked into a separate repo.

Updates momento dep to the latest version to get all the currently supported sorted set implementations.

Adds most of the sorted set commands to the momento proxy:

- ZCARD <-> SortedSetLength
- ZINCRBY <-> SortedSetIncrement
- ZSCORE <-> SortedSetGetScore
- ZMSCORE <-> SortedSetGetScores
- ZREM <-> SortedSetRemoveElements
- ZRANK <-> SortedSetGetRank (lowest score rank 0)
- ZRANGE <-> SortedSetFetchByRank, SortedSetFetchByScore -- Momento does not appear to support the BYLEX option so that is also unsupported in the proxy
- ZADD <-> SortedSetPutElements -- Momento does not appear to support many of the optional arguments so they are also unsupported in the proxy. Only INCR was easily achievable so it's supported.
- ZREVRANK <-> SortedSetGetRank (highest score rank 0)
- ZCOUNT <-> SortedSetLengthByScore 
- ZUNIONSTORE <-> SortedSetUnionStore

Been testing using unit tests as well as redis cli commands

<details>

<summary>Sample CLI commands</summary>

Basic get and set:
```
redis-cli -h 0.0.0.0 -p 6379 GET FOO
redis-cli -h 0.0.0.0 -p 6379 SET FOO BAR
redis-cli -h 0.0.0.0 -p 6379 GET FOO
```

Sorted set length:
```
redis-cli -h 0.0.0.0 -p 6379 ZCARD myzset
```

Sorted set count (length by score):
```
redis-cli -h 0.0.0.0 -p 6379 ZCOUNT myzset
redis-cli -h 0.0.0.0 -p 6379 ZCOUNT myzset 1 100
redis-cli -h 0.0.0.0 -p 6379 ZCOUNT myzset "(1" 10
redis-cli -h 0.0.0.0 -p 6379 ZCOUNT myzset -inf +inf
redis-cli -h 0.0.0.0 -p 6379 ZCOUNT myzset +inf -inf
```

Sorted set increment:
```
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 1 "one"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 2 "two"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 3 "three"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 4 "four"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 5 "five"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset 6 "six"
redis-cli -h 0.0.0.0 -p 6379 ZINCRBY myzset +inf "inf"
```

Sorted set get score:
```
redis-cli -h 0.0.0.0 -p 6379 ZSCORE myzset "one"
```

Sorted set get multiple scores:
```
redis-cli -h 0.0.0.0 -p 6379 ZMSCORE myzset "one" "two" "abc"
redis-cli -h 0.0.0.0 -p 6379 ZMSCORE notaset "one" "two" "three"
redis-cli -h 0.0.0.0 -p 6379 ZMSCORE myzset "a" "b" "c" "abc"
```

Sorted set remove elements:
```
redis-cli -h 0.0.0.0 -p 6379 ZREM notaset "one"
redis-cli -h 0.0.0.0 -p 6379 ZREM myzset "one"
```

Sorted set get rank (ascending order, scores low to high):
```
redis-cli -h 0.0.0.0 -p 6379 ZRANK notaset "one"
redis-cli -h 0.0.0.0 -p 6379 ZRANK myzset "one"
redis-cli -h 0.0.0.0 -p 6379 ZRANK myzset "one" WITHSCORE
redis-cli -h 0.0.0.0 -p 6379 ZRANK myzset "two"
redis-cli -h 0.0.0.0 -p 6379 ZRANK myzset "abc"
```

Sorted set reverse rank (descending order, scores high to low):
```
redis-cli -h 0.0.0.0 -p 6379 ZREVRANK notaset "one"
redis-cli -h 0.0.0.0 -p 6379 ZREVRANK myzset "one"
redis-cli -h 0.0.0.0 -p 6379 ZREVRANK myzset "one" WITHSCORE
redis-cli -h 0.0.0.0 -p 6379 ZREVRANK myzset "two"
redis-cli -h 0.0.0.0 -p 6379 ZREVRANK myzset "abc"
```

Sorted set fetch by rank:
```
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 -10
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 REV
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 REV WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 LIMIT 1 1 -- does not work as expected
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 BYLEX -- does not work as expected
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 "(10" -- does not work as expected
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -10 +10 -- works but +10 maybe shouldn't? 
```

Sorted set fetch by score:
```
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 0 10 BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -inf 10 BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -inf +inf BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset +inf -inf BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset 5 +inf BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -inf +inf BYSCORE WITHSCORES
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -inf +inf BYSCORE WITHSCORES REV
redis-cli -h 0.0.0.0 -p 6379 ZRANGE myzset -inf +inf BYSCORE WITHSCORES REV LIMIT 1 5
```

Sorted set put elements:
```
redis-cli -h 0.0.0.0 -p 6379 ZADD myzset -inf "negative-infinity" +inf "positive-infinity"
redis-cli -h 0.0.0.0 -p 6379 ZADD myzset INCR 1.23 "a-float"
redis-cli -h 0.0.0.0 -p 6379 ZADD myzset 1.0 "a" 2.34 "b" 567 "c"
```

Sorted set union store:
```
# Preparation
redis-cli -h 0.0.0.0 -p 6379 ZADD zset1 1 "one"
redis-cli -h 0.0.0.0 -p 6379 ZADD zset1 2 "two"
redis-cli -h 0.0.0.0 -p 6379 ZADD zset2 1 "one"
redis-cli -h 0.0.0.0 -p 6379 ZADD zset2 2 "two"
redis-cli -h 0.0.0.0 -p 6379 ZADD zset2 3 "three"

# Actual union attempts
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 0 -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 1 myzset -- should work
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE MAX
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE MIN
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE SUM
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 MAX AGGREGATE -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE MAX WEIGHTS -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS AGGREGATE MAX -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS 1 2.3
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS 1 -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS 1 2 3 -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE MAX WEIGHTS 1 2
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 AGGREGATE MAX WEIGHTS 1 2 3 -- should error
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS 1 2 AGGREGATE MIN
redis-cli -h 0.0.0.0 -p 6379 ZUNIONSTORE dest 2 zset1 zset2 WEIGHTS 1 2 3 AGGREGATE MIN -- should error
```

</details>

brayniac reviewed (changes requested):
Looks good overall. Let's address the clippy warnings, change the year in the copyright headers for new files, and make sure that Momento specifics are contained in the proxy code and not in the protocol itself.

anitarua commented:
@brayniac I've reworked mostly ZADD and ZRANGE to address your feedback, as well as the clippy warnings and copyright statements

malandis commented on src/protocol/resp/src/request/smembers.rs:
```diff
@@ -53,7 +53,7 @@ impl From<&SetMembers> for Message {
         Message::Array(Array {
             inner: Some(vec![
-                Message::BulkString(BulkString::new(b"GET")),
+                Message::BulkString(BulkString::new(b"SMEMBERS")),
```
Nice find.

malandis commented on src/proxy/momento/src/protocol/resp/zadd.rs:
```diff
@@ -0,0 +1,96 @@
+
+        // If INCR is set, then ZADD should behave like ZINCRBY (as per the docs), which accepts only a single score-member pair
+        if req.optional_args().incr {
+            if req.members().len() != 1 {
+                klog_1(&"zadd", &req.key(), Status::ServerError, 0);
+                return Err(ProxyError::from(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "INCR option requires exactly one score-member pair",
+                )));
```
This logic should be in the request parsing (ie any request like this should be malformed, whether it is going to a Momento backend or any other backend).

malandis commented on src/proxy/momento/src/protocol/resp/zcount.rs:
```diff
@@ -0,0 +1,68 @@
+    req: &SortedSetCount,
+) -> ProxyResult {
+    update_method_metrics(&ZCOUNT, &ZCOUNT_EX, async move {
+        let (min_score, min_score_inclusive) = parse_score_boundary_as_float(req.min_score())?;
+        let (max_score, max_score_inclusive) = parse_score_boundary_as_float(req.max_score())?;
```
This belongs in the request parsing since the scores have a special syntax. When that syntax is invalid, the request is invalid and should be reject at that level (here and elsewhere).

That way the code here can access `min_score` and `min_score_inclusive` from the parsed request object.

malandis commented on src/proxy/momento/src/protocol/resp/zcount.rs:
```diff
@@ -0,0 +1,68 @@
+        let (min_score, min_score_inclusive) = parse_score_boundary_as_float(req.min_score())?;
+        let (max_score, max_score_inclusive) = parse_score_boundary_as_float(req.max_score())?;
+        let request = SortedSetLengthByScoreRequest::new(cache_name, req.key())
+            .min_score(if min_score_inclusive {
+                min_score
+            } else {
+                min_score + 1.0
+            })
+            .max_score(if max_score_inclusive {
+                max_score
+            } else {
+                max_score - 1.0
+            });
```
Also need to check for unbounded scores.

malandis commented on src/protocol/resp/src/request/zcount.rs:
```diff
@@ -0,0 +1,141 @@
+                .parse(b"ZCOUNT z -inf +inf\r\n")
+                .unwrap()
+                .into_inner(),
+            Request::SortedSetCount(SortedSetCount::new(b"z", b"-inf", b"+inf"))
+        );
+
+        assert_eq!(
+            parser.parse(b"ZCOUNT z (1 3\r\n").unwrap().into_inner(),
+            Request::SortedSetCount(SortedSetCount::new(b"z", b"(1", b"3"))
+        );
```
Should parse out the score boundaries in this code as well, since an invalid score should be rejected.

malandis commented on src/protocol/resp/src/request/zincrby.rs:
```diff
@@ -0,0 +1,132 @@
+            Request::SortedSetIncrement(SortedSetIncrement::new(b"z", b"1", b"a"))
+        );
+
+        assert_eq!(
+            parser.parse(b"ZINCRBY z +inf a\r\n").unwrap().into_inner(),
+            Request::SortedSetIncrement(SortedSetIncrement::new(b"z", b"+inf", b"a"))
+        );
+
```
Should parse out the numbers since an invalid request should be rejected here.

anitarua replied:
Fyi, parsing out the float and storing it in the `SortedSetIncrement` struct will mean the `Eq` trait is no longer satisfied. Removing `Eq` then requires removing it from [here](https://github.com/pelikan-io/pelikan/blob/main/src/protocol/resp/src/request/mod.rs#L101), which I think is okay, just means a less strict equivalence constraint

malandis reviewed (commented):
A cross cutting change I recommend is to handle parsing and validating score boundaries in the request parser as opposed to at the Momento translation level for two reasons: (1) a malformed score is an invalid RESP request, not an invalid translation to Momento; (2) the code to build the Momento requests will be simpler when reading from fully parsed scores already.

anitarua commented:
@malandis @brayniac 
I've updated the RESP request parsing with the cross-cutting changes that were recommended. 

I've also split out the momento-proxy changes into a new PR on the momento fork of the proxy where we'll continue development: https://github.com/momentohq/momento-proxy/pull/2
This PR will need to upgrade the `protocol-resp` package version once this PR is merged.

brayniac reviewed (approved):
LGTM
